### ISSUE TITLE
Add SARIF renderer

### ISSUE THREAD

**Issue by lukasbestle** [phpmd/phpmd#858]
## Description

[SARIF, the Static Analysis Results Interchange Format](https://github.com/microsoft/sarif-tutorials/blob/main/docs/1-Introduction.md), is a standard, JSON-based format for the output of static analysis tools.

As it is an approved OASIS standard, many other tools already support it, including other analyzers and also visualizers (for example IDEs or [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/github/finding-security-vulnerabilities-and-errors-in-your-code/sarif-support-for-code-scanning)). SARIF support in PHPMD would therefore make it much more interoperable with other tools and workflows.

Because SARIF is a JSON-based format, adding support for it to PHPMD is quite simple. If other community members and maintainers agree that it is useful, I can write the code for it. But I'd wait until my other PR #857 is merged to avoid merge conflicts and also so code review feedback for that other renderer can be incorporated directly when building the SARIF renderer.

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] Tell if you have the option to provide the code for this proposal.
* [x] This issue is about 1 feature proposal and nothing more.
* [x] The issue has a descriptive title. For example: "Add JSON render option".

**Comment by lukasbestle** [comment]
@kylekatarnls @MarkVaughn Would you merge a PR for this? I think it would work quite well in combination with the GitHub renderer (I plan to use both together in my projects).

**Comment by kylekatarnls** [comment]
Hi @lukasbestle by any change, is it the format GitLab uses?
https://docs.gitlab.com/ee/user/project/merge_requests/code_quality.html#implementing-a-custom-tool

**Comment by lukasbestle** [comment]
@kylekatarnls No, that expects the Code Climate spec, which is not the same as SARIF. SARIF is an official standard while Code Climate is a format of a particular set of tools.

**Comment by kylekatarnls** [comment]
This make me think that the list of rendered could grow very fast if we implement every formats. We rather need to have mapping options.

We have:
#856 which is XML with a custom mapping
SARIF and CodeClimate/GitLab which is JSON with a custom mapping
And maybe #857 could also be use the text with a custom settings (template, prefixes/suffixes)

If we add them all as first-level renderers, we'll have 9 of them. Let's say it would be the very maximum before considering reorganizing this feature.

**Comment by lukasbestle** [comment]
I see your point, this is definitely something to be considered.

The question is how high the abstraction level should be. I'm generally not a huge fan of abstractions in areas like these as it will make the code much harder to understand, longer than necessary and each change in the "core format" (XML, JSON, text) might break the formats that depend on them.

SARIF is quite specific, it's not just JSON. For example it also includes the rule metadata (description, help text, example, link to the docs etc.). So basically the only thing that would be shared is the `json_encode()` at the end.

Another argument for keeping a linear list instead of thinking in base formats: As a user I might not even know if I'm looking for a JSON-based format or an XML format. I know that I need SARIF or Checkstyle etc. How these formats work internally is not something I'd consider as a user, it's an implementation detail.

What we could do is to move the renderer registration into a central location. So instead of a method for each renderer and an additional line in the command options for file export, there could be a static array that contains a list of renderer names and classes, which would be easier to maintain and less verbose:

```php
const RENDERERS = [
  'text' => [
    'class'     => '\PHPMD\Renderer\TextRenderer',
    'forExport' => false 
  ],
  'sarif' => [
    'class'     => '\PHPMD\Renderer\SarifRenderer',
    'forExport' => true 
  ],
  // ...
];
```

Alternatively it could also be a `public static` so that plugins can register themselves as well.

The `forExport` attribute would control whether the renderer gets its own `--reportfile-{name}` option.

**Comment by ffflabs** [comment]
I kind of agree with @kylekatarnls in that stuffing the actual library with so many report renderers becomes more and more confusing for users and harder to mantain.

On the other hand, the current approach to use custom renderers isn't really straightforward regarding how should one address the custom renderers class, if it means a path, a fqcn, and if backslashes in its namespace should be escaped.

Perhaps it would be saner to allow for the cli to pass a config file

**Comment by lukasbestle** [comment]
I agree that a config file could be useful in general. However I don‘t think that a config file alone can solve the issue of how to get custom renderers loaded. If you declare a class in the config file, PHPMD still needs to know how to load that class.

To be honest I think the ideas we collected in this issue are mostly medium or long term improvements. I don‘t think they should be blockers for new core renderers that are generally useful and not too project-specific.

SARIF as an open and official standard (*the* format for static analyzer output) should IMO be in the core.

**Comment by ffflabs** [comment]
> I agree that a config file could be useful in general. However I don‘t think that a config file alone can solve the issue of how to get custom renderers loaded. If you declare a class in the config file, PHPMD still needs to know how to load that class.
> 

As long as the custom renderer keeps the interface contract (if there was one) PHPMD doesn't need to know anyting. In fact, it's working that way right now. I've set up a POC using nothing but 

 - vanilla PHPMD as of aa7e527b5068, 
 - your GithubRenderer #857 
 - [a Github Action with 4 scenarios](https://github.com/ffflabs/phpmd_custom_renderer_poc/pull/1/checks)
    1.- binary at vendor/bin/phpmd, GitHubRenderer on project root, no namespace
    2.- same as above with globally installed phpmd
    3.- same as 1, but having GithubRenderer on a subfolder and with a namespace
    4. same as 3 with global phpmd binary
 
It turns out that referencing the custom renderer class in project root and no namespace **always works**. Namespaced custom renderers will work with _local_ binary, but not with _global_ binary. Why? it is because class existance takes the autoloader in consideration, therefore class does exist.

https://github.com/phpmd/phpmd/blob/aa7e527b50681ca4be0f5457df6a4de5fdabe52c/src/main/php/PHPMD/TextUI/CommandLineOptions.php#L455-L457

If it doesn't, including the file doesn't work, because it either gets the path right (but the FQCN wrong) or the other way around. It cannot get both right since method  `createCustomRenderer` assumes the path is just the FQCN with forward slashes relative to `cwd`

https://github.com/phpmd/phpmd/blob/aa7e527b50681ca4be0f5457df6a4de5fdabe52c/src/main/php/PHPMD/TextUI/CommandLineOptions.php#L459-L470

So I wonder if this all could be tackled by adding a config flag like `--custom-renderers-folder=<path>` so

```bash
phpmd src/php/PHPMD  Whatever\Namespace\CustomRenderer cleancode --custom-renderers-folder=/whatever/absolute/or/relative/path
```



> To be honest I think the ideas we collected in this issue are mostly medium or long term improvements. I don‘t think they should be blockers for new core renderers that are generally useful and not too project-specific.
> 
> SARIF as an open and official standard (_the_ format for static analyzer output) should IMO be in the core.

I didn't know about SARIF, but I can see now it's a broadly used standard.

**Comment by kylekatarnls** [comment]
Fixed via #865
